### ISSUE TITLE
udev rules may get missed due to rule caching

### ISSUE THREAD

**Issue by khautio** [systemd/systemd#3134]
### Submission type
- Bug report
### systemd version the issue has been seen with

Tested with 229
### Used distribution

RHEL, Fedora, LFS
### In case of bug report: Expected behaviour you didn't see

Newly added rules are immediately effective
### In case of bug report: Unexpected behaviour you saw

Rules may get ignored due to caching
### In case of bug report: Steps to reproduce the problem

``` sh
#!/bin/bash                                                                                                                                                                              

chown root.root /dev/zero
echo "KERNEL==\"zero\", GROUP=\"nobody\"" > 80-zero.rules

udevadm trigger --sysname-match=zero --action=add
cp -v 80-zero.rules /usr/lib/udev/rules.d/
test "$1" && udevadm control --reload
udevadm trigger --sysname-match=zero --action=add
ls -l /dev/zero
sleep 1
udevadm trigger --sysname-match=zero --action=add
ls -l /dev/zero
sleep 1
udevadm trigger --sysname-match=zero --action=add
ls -l /dev/zero
sleep 1
udevadm trigger --sysname-match=zero --action=add
ls -l /dev/zero
sleep 1
udevadm trigger --sysname-match=zero --action=add
ls -l /dev/zero
sleep 1
udevadm trigger --sysname-match=zero --action=add
ls -l /dev/zero
sleep 1
udevadm trigger --sysname-match=zero --action=add
ls -l /dev/zero
```

```
[root@kha foo2]# ./utest.sh
‘80-zero.rules’ -> ‘/usr/lib/udev/rules.d/80-zero.rules’
crw-rw-rw-. 1 root root 1, 5 27. 4. 13:39 /dev/zero
crw-rw-rw-. 1 root root 1, 5 27. 4. 13:39 /dev/zero
crw-rw-rw-. 1 root root 1, 5 27. 4. 13:39 /dev/zero
crw-rw-rw-. 1 root root 1, 5 27. 4. 13:39 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:39 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:39 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:39 /dev/zero
[root@kha foo2]# 
```

**Comment by khautio** [comment]
Forgot to add the run with parameter (calling reload in between)

```
[root@kha foo2]# ./utest.sh 1
‘80-zero.rules’ -> ‘/usr/lib/udev/rules.d/80-zero.rules’
crw-rw-rw-. 1 root root 1, 5 27. 4. 13:40 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:40 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:40 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:40 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:40 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:40 /dev/zero
crw-rw-rw-. 1 root nobody 1, 5 27. 4. 13:40 /dev/zero
[root@kha foo2]# 
```

**Comment by khautio** [comment]
This bug was first noticed when installing qemu-kvm rpm on RHEL system. Said rpm includes an udev rule file to set the group and permissions of /dev/kvm and triggers udev in its post script, depending on timing this may fail.

**Comment by khautio** [comment]
I just realised I should call udev settle after trigger to wait actions to complete.  I am quite sure the case with reload will then work.

**Comment by khautio** [comment]
I just realised I should call udev settle after trigger to wait actions to complete.  I am quite sure the case with reload will then work.

**Comment by khautio** [comment]
The issue is not to be closed, it still fails if "udev control --reload" is not done.Calling reload should not  be needed as udev is supposed to monitor it's rule directories.

revised testcase 

```
root@distro:~$ systemctl --version
systemd 229
+PAM -AUDIT -SELINUX -IMA -APPARMOR -SMACK +SYSVINIT +UTMP -LIBCRYPTSETUP -GCRYPT -GNUTLS +ACL +XZ +LZ4 -SECCOMP +BLKID +ELFUTILS +KMOD +IDN
root@distro:~$ cat utest.sh
#!/bin/bash

rm -f /usr/lib/udev/rules.d/80-foo.rules  /usr/lib/udev/rules.d/80-zero.rules
chown root.root /dev/zero
echo "KERNEL==\"zero\", GROUP=\"nobody\"" > 80-zero.rules

udevadm trigger --sysname-match=null --action=add
cp -v 80-zero.rules /usr/lib/udev/rules.d/
for ((N=0; N<5; N++)) ; do
  udevadm trigger --sysname-match=zero --action=add
  udevadm settle
  ls -l /dev/zero
  sleep 1
done
root@distro:~$ ./utest.sh
'80-zero.rules' -> '/usr/lib/udev/rules.d/80-zero.rules'
crw-rw-rw- 1 root root 1, 5 Apr 28 06:29 /dev/zero
crw-rw-rw- 1 root root 1, 5 Apr 28 06:29 /dev/zero
crw-rw-rw- 1 root root 1, 5 Apr 28 06:29 /dev/zero
crw-rw-rw- 1 root nobody 1, 5 Apr 28 06:29 /dev/zero
crw-rw-rw- 1 root nobody 1, 5 Apr 28 06:29 /dev/zero
root@distro:~$ 
```

**Comment by evverx** [comment]
> Calling reload should not be needed as udev is supposed to monitor it's rule directories

``` c
/* check for changed config, every 3 seconds at most */
if (manager->last_usec == 0 ||
        (usec - manager->last_usec) > 3 * USEC_PER_SEC) {
          if (udev_rules_check_timestamp(manager->rules) ||
                 udev_builtin_validate(manager->udev))
                        manager_reload(manager);
```

https://github.com/systemd/systemd/blob/master/src/udev/udevd.c#L796

Try:

```
cp -v 80-zero.rules /usr/lib/udev/rules.d/
+ sleep 3
```

Why not just use `udevadm control --reload`?

**Comment by khautio** [comment]
Yeah sleeping for 3 seconds or calling reload will both work. I guess this is then a feature of systemd/udev.

If seen as normal functionality then all users should be changed to cope with this; basically by adding reload before triggering rules.

**Comment by poettering** [comment]
Yeah, if you want to settle, you need "udevadm setlle", and if you want to reload, use "udevadm control --reload". Hence I think there's no bug, but just intended behaviour. Closing hence.

**Comment by dmick** [comment]
Why doesn't systemd/udev use inotify to monitor its dirs rather than polling, anyway?

**Comment by keszybz** [comment]
Current solution is semi-useless: because of the 3 second timeout, one cannot depend on the rules being reloaded. So either one needs to force the reload (when the updated rules are necessary before proceeding) and the automatic reload is useless, or one doesn't need the updated rules, so the automatic reload is useless.

I think we should remove the 3s timeout, and *always* reload, before handling an event, after `udev_rules_check_timestamp()` returns true. The only time when this would be a problem would be if somebody was maliciously modifying timestamps and triggering events in parallel. We can add some throttle for that, but I don't even think that it's necessary.

**Comment by evverx** [comment]
Hm, I'm reading http://git.kernel.org/cgit/linux/hotplug/udev.git

[2005-10-27 19:04:38 (GMT)](http://git.kernel.org/cgit/linux/hotplug/udev.git/commit/?id=c895fd002763d6ae808e8)
> If inotify is available, udevd will reload all rules if any change in
/etc/udev/rules.d/ happens. Otherwise -HUP or "udevcontrol reload_rules"
can be used.

[2011-12-27 02:49:43 (GMT)](http://git.kernel.org/cgit/linux/hotplug/udev.git/commit/?id=4f1795cca7060030461c8f3)
> replace inotify with time-controlled stat()

So, actually, this issue is a regression

**Comment by evverx** [comment]
> Why doesn't systemd/udev use inotify to monitor its dirs rather than polling, anyway?

I think the main reason is
> Inotify reports only events that a user-space program triggers
       through the filesystem API.  As a result, it does not catch remote
       events that occur on network filesystems.  (Applications must fall
       back to polling the filesystem to catch such events.)

But I'm not sure

**Comment by evverx** [comment]
@keszybz ,
> and always reload, before handling an event

every event?
I didn't measure of course. But seems like this will slow down the `udevd` heavily

**Comment by keszybz** [comment]
I think we can assume that udev rules are not loaded from a network fs. And even if they are, I think we can assume that the fs is not changing underneath us.

> every event?

Only if we detect a change. So that's mean three extra `stat` calls or so, from a hot cache. I'm pretty sure that's immeasurable.

**Comment by evverx** [comment]
@keszybz , why not to mark as the "regression"?

**Comment by evverx** [comment]
> I think we can assume that udev rules are not loaded from a network fs.

It doesn't really matter but we can't assume that. Some people [mount the root filesystem via NFS](https://www.kernel.org/doc/Documentation/filesystems/nfs/nfsroot.txt).

> I'm pretty sure that's immeasurable

Well, I'm not sure about `t(three extra stat calls)` * huge number of events. Anyway, why not to measure? :-)

**Comment by yuwata** [comment]
So, should this be fixed? systemd or systemd daemons do not reload config files automatically, and we need to run, e.g. `systemctl daemon-reload` or `systemctl restart systemd-networkd.service`, and there exists `udevadm control --reload`.
